lib/logical_solver/presolve.hpp

```diff
@@ -0,0 +1,455 @@
+
+  // PUBLIC METHODS
+  /**
+   * @brief Applies presolve to the problem.
+   * 
+   * @return void.
+   */
+  void applyPresolve();
+
+  /**
+   * @brief Applies postsolve to the problem.
+   * 
+   * @return void.
+   */
+  void applyPostsolve();
+
+  /**
+   * @brief Prints the feasible solutions vector.
+   */
+  void printFeasibleSolution();
+
+  /**
+   * @brief Prints the instance LP.
+   */
+  void printLP();
+
+  /**
+   * @brief Prints the constraints bounds.
+   */
+  void printBounds();
+
+  /**
+   * @brief Prints the implied bounds.
+   */
+  void printImpliedBounds();
+
+  private:
+  // PRIVATE INSTANCE VARIABLES 
+  int kInfinity;
+
+  // vectors to keep track of active rows and columns
+  // during presolve
+  std::vector<bool> presolve_active_rows_;
+  std::vector<bool> presolve_active_columns_;
+  int presolve_active_rows_count_;
+  int presolve_active_columns_count;
+
+  // Vector to check we don't apply inequality row 
+  // singletons presolve twice to the same inequality.
+  std::vector<int> inequality_singletons_;
+
+  std::vector<bool> postsolve_active_rows_;
+  std::vector<bool> postsolve_active_cols_;
+
+  // Array to keep track of the non-zero active variables in each
+  // row and col during presolve.
+  std::vector<std::vector<int>> rows_non_zero_variables_;
+  std::vector<std::vector<int>> cols_non_zeros_indices_;
+
+  // STRUCTS:
+  // struct containing the search function and update 
+  // state function for a given rule (pointers).
+
+  // (type definition for trying to add postsolve function
+  // to the stack)
+  // typedef void (Presolve::*member_function)(int, int);
+
+  // struct to keep track of presolve rules applied 
+  // during presolve.
+  struct presolve_log {   
+    int constraint_index;      // constraint number      
+    int variable_index;        // variable (column) index
+    int rule_id; // presolve rule function to apply in postsolve
+    std::vector<int> dependancies;
+  }; 
+
+  // PRIVATE CLASS VARIABLES:
+  // stack of presolve structs.
+  std::stack<presolve_log> presolve_stack_;
+
+  // PRIVATE METHODS
+  /**
+   * @brief Gets the indices of the non-zero columns
+   * of each row (non-zero variables) and stores in the 
+   * instance variable rows_non_zero_variables.
+   * 
+   * @return void
+   */
+  void getRowsNonZeros();
+
+  /**
+   * @brief Gets the indices of the non-zero rows (coefficients)
+   * of each column and stores in the 
+   * instance variable cols_non_zeros_indices.
+   * 
+   * @return void
+   */
+  void getColsNonZeros();
+
+  /**
+   * @brief Gets the indices of the non-zero rows (coefficients)
+   * of each column and stores in the 
+   * instance variable cols_non_zeros_indices;
+   * and same for the non-zero variables in each row, storing them
+   * in the variable rows_non_zero_variables.
+   * 
+   * @return void
+   */
+  void getRowsAndColsNonZeros();
+
+  double getFeasibleValueCalculationBound(int row_index);
+
+  bool checkIsRowFree(int row_index);
```

Comment:
Function descriptions here!

Reply:
I know it seems pointless for functions like this but it's worth it for someone (such as a marker) to see it...

Reply:
Yeah need to add them.